Prepare release notes for the upcoming v0.2.0

joseph-wakeling-sociomantic opened the pull request:
These patches add a `Build/Use` section to the `README` file that covers dependencies (including submodule versions), and adds Neptune-style release notes for all changes made since the v0.1.0 release.

joseph-wakeling-sociomantic committed:
Add Build/Use, Releases, and Contributing sections to README file

The `Build/Use` section describes the dependencies of the library, with
package and submodule dependencies covered in separate subsections.
Notes on how to build, run tests, and convert to D2 are also provided
(the latter replacing the original section on D2 compatibility).

The `Releases` and `Contributing` sections are standardized sections for
Neptune libraries: the first describes where to find library releases
and the structure of the release notes, while the second provides a link
to contributor documentation (which includes release-notes management).

joseph-wakeling-sociomantic committed:
Add release notes for changes made since v0.1.0

This should make it easier to generate nice release notes for the v0.2.0
release.  While a full release-notes procedure might be overkill while
we're still in the v0.x.x release stage, it's never too early to adopt
this practice.

joseph-wakeling-sociomantic commented:
Question: should I still include an explicit `RELEASE_NOTES.md` file (containing the instructions on how release notes are prepared) or is this something we can avoid?

jens-mueller-sociomantic commented:
Can you point to existing documentation?

gavin-norman-sociomantic commented:
>Question: should I still include an explicit RELEASE_NOTES.md file (containing the instructions on how release notes are prepared) or is this something we can avoid?

In other libraries, that file just describes the "new" release notes system. I'd suggest that this info be moved to the README, though. I'll make a PR in ocean for this and then (once it's reviewed and finalised) maybe you could follow the same procedure here.

gavin-norman-sociomantic commented:
See https://github.com/sociomantic-tsunami/neptune/pull/117  and https://github.com/sociomantic-tsunami/ocean/pull/337.

joseph-wakeling-sociomantic commented:
> Can you point to existing documentation?

What documentation did you have in mind?

joseph-wakeling-sociomantic commented:
Updated the `README` in line with @gavin-norman-sociomantic's ocean PR.

jens-mueller-sociomantic commented:
> What documentation did you have in mind?

If there was documentation on how to prepare release notes, I'd prefer a link. This assumes that projects can share such documentation. I thought neptune pinned down how to write such release notes. I only find https://github.com/sociomantic-tsunami/neptune/blob/master/doc/announcements.md.

gavin-norman-sociomantic commented:
I've just made a PR to add this doc to neptune today: sociomantic-tsunami/neptune#117.

joseph-wakeling-sociomantic commented:
> If there was documentation on how to prepare release notes, I'd prefer a link.

Ah, good.  The `squash!` commit adds this.

joseph-wakeling-sociomantic commented on relnotes/handle.feature.md:
```diff
@@ -0,0 +1,12 @@
+  ``freeHandle`` method.  This can be used to track down potential
+  memory leaks in an application using dmxnet.
+
+  A new private ``debugReport`` function has been added to provide
```
Wondering whether this should be mentioned at all, but it's useful at least for developers to know about.

joseph-wakeling-sociomantic commented:
Are there any objections to this PR as it is?  Would quite like to expedite the new release if possible.

jens-mueller-sociomantic commented:
Can you squash?

joseph-wakeling-sociomantic commented:
@jens-mueller-sociomantic I was asking about content ;-)

Squashed, anyway, and updated the 1st commit title & message to better describe the changes.

jens-mueller-sociomantic commented:
Are the file names like `ndarray.feature.md` too generic? I mean if we release another feature for the `NDArray` what will the file name of that change.

joseph-wakeling-sociomantic commented:
@mihails-strasuns-sociomantic @gavin-norman-sociomantic can you advise on @jens-mueller-sociomantic's question?  I think that for the current use-case things are fine but we might want to adopt a different rule as things go on.

In particular, I'm worried about multiple repeated bullet points for the same module (instead of one bullet point per module, with multiple new features listed in there).

mihails-strasuns-sociomantic commented:
Currently we name files by features, not modules. Same goes for resulting release notes - they are grouped by distinct features (that may cover one or many modules) and if one modules is affected by several different feature implementations, it is appropriate for it to appear in the changelog several times.

joseph-wakeling-sociomantic commented:
OK, I'll edit accordingly.

joseph-wakeling-sociomantic commented:
Updated second patch to split into feature-by-feature notes.  The `ndarray.feature.md` file has been left as is since it affects the `NDArray` class.

jens-mueller-sociomantic commented:
Why not split into `ndarray-improve-doc-opSubAssign.feature.md` and `ndarray-improve-doc-opMulAssign.feature.md`? I think `ndarray.feature.md` is to generic as a file name. Keeping it as one feature is also fine, but then something like `ndarray-improve-doc.feature.md` might be better.

I also prefer `softmaxoutput-symbol-doc.feature.md` over `softmaxoutput-symbol.feature.md` since it is less general.

joseph-wakeling-sociomantic commented:
> Why not split into ndarray-improve-doc-opSubAssign.feature.md and ndarray-improve-doc-opMulAssign.feature.md?

Well, I can do it if you really feel it's necessary.  It just seems a bit overkill for this particular use-case.

I'm fine to make the filename `ndarray-improve-doc` though.

joseph-wakeling-sociomantic commented:
Some file renames and one small content tweak:

```diff
diff --git a/relnotes/ndarray.feature.md b/relnotes/ndarray-doc.feature.md
similarity index 79%
rename from relnotes/ndarray.feature.md
rename to relnotes/ndarray-doc.feature.md
index 701152e..83f435c 100644
--- a/relnotes/ndarray.feature.md
+++ b/relnotes/ndarray-doc.feature.md
@@ -1,4 +1,4 @@
-* ``mxnet.NDArray``
+* ``mxnet.NDArray.NDArray``
 
   Some small documentation improvements have been added to the
   `opSubAssign` and `opMulAssign` methods.
diff --git a/relnotes/softmaxoutput-symbol.feature.md b/relnotes/softmaxoutput-symbol-doc.feature.md
similarity index 100%
rename from relnotes/softmaxoutput-symbol.feature.md
rename to relnotes/softmaxoutput-symbol-doc.feature.md
```

jens-mueller-sociomantic commented:
LGTM.

joseph-wakeling-sociomantic commented:
Any other feedback @mihails-strasuns-sociomantic and @gavin-norman-sociomantic ... ?

mihails-strasuns-sociomantic commented on README.rst:
```diff
@@ -18,14 +18,56 @@ Further design constraints include:
+========== =======
+
+``beaver`` is only used by the Travis CI setup, so can be ignored for general
+development or use of the library.
```
I would probably remove specific version from dependency table as it doesn't matter for users of the lib.

joseph-wakeling-sociomantic replied:
Yea, I note that swarm does not list it at all.  Maybe better to remove it from the table entirely and mention it only below?

mihails-strasuns-sociomantic replied:
Makes sense to me

joseph-wakeling-sociomantic replied:
I mean, if we're going to have it in the table, I'd rather list the version (and find some way to make clear this is a CI dependency, not a library dependency) than be inconsistent.

mihails-strasuns-sociomantic reviewed (approved):
Just one tiny nitpick

joseph-wakeling-sociomantic commented:
Pushed a fixup patch for the `beaver` submodule dependency.  @mihails-strasuns-sociomantic LGTY?

mihails-strasuns-sociomantic commented:
Yup

joseph-wakeling-sociomantic commented:
OK, squashed down.  Taking this in as soon as CI agrees.

joseph-wakeling-sociomantic commented:
(Not that there are any changes that could trigger CI failure, but on principle....)
